src/script/src/python/ffi_types/vector.rs

```diff
@@ -0,0 +1,524 @@
+                value::Value::UInt64(v) => Box::new(Int64Array::from_value(v as i64, left_len)),
+                value::Value::Float64(v) => Box::new(Int64Array::from_value(v.0 as i64, left_len)),
+                _ => unreachable!(),
+            }
+        } else if is_unsigned(&target_type) {
+            match right {
+                value::Value::Int64(v) => Box::new(UInt64Array::from_value(v as u64, left_len)),
+                value::Value::UInt64(v) => Box::new(UInt64Array::from_value(v, left_len)),
+                value::Value::Float64(v) => Box::new(UInt64Array::from_value(v.0 as u64, left_len)),
+                _ => unreachable!(),
+            }
+        } else {
+            return Err(format!(
+                "Can't cast source operand of type {:?} into target type of {:?}",
+                right_type, &target_type
+            ));
+        };
+
+        let result = op(left.as_ref(), right.as_ref())?;
+
+        Ok(Helper::try_into_vector(result.clone())
+            .map_err(|e| format!("Can't cast result into vector, result: {result:?}, err: {e:?}",))?
+            .into())
+    }
+
+    pub(crate) fn rspy_scalar_arith_op<F>(
+        &self,
+        other: PyObjectRef,
+        target_type: Option<ArrowDataType>,
+        op: F,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyVector>
+    where
+        F: Fn(&dyn Array, &dyn Array) -> Result<ArrayRef, String>,
+    {
+        // the right operand only support PyInt or PyFloat,
+        let right = {
+            if is_instance::<PyInt>(&other, vm) {
+                other.try_into_value::<i64>(vm).map(value::Value::Int64)?
+            } else if is_instance::<PyFloat>(&other, vm) {
+                other
+                    .try_into_value::<f64>(vm)
+                    .map(|v| (value::Value::Float64(OrderedFloat(v))))?
+            } else {
+                return Err(vm.new_type_error(format!(
+                    "Can't cast right operand into Scalar of Int or Float, actual: {}",
+                    other.class().name()
+                )));
+            }
+        };
+        self.scalar_arith_op(right, target_type, op)
+            .map_err(to_type_error(vm))
+    }
+
+    pub(crate) fn vector_arith_op<F>(
+        &self,
+        right: &Self,
+        target_type: Option<ArrowDataType>,
+        op: F,
+    ) -> Result<PyVector, String>
+    where
+        F: Fn(&dyn Array, &dyn Array) -> Result<ArrayRef, String>,
+    {
+        let left = self.to_arrow_array();
+        let right = right.to_arrow_array();
+
+        let left_type = &left.data_type();
+        let right_type = &right.data_type();
+
+        let target_type = target_type.unwrap_or_else(|| {
+            if is_signed(left_type) && is_signed(right_type) {
+                ArrowDataType::Int64
+            } else if is_unsigned(left_type) && is_unsigned(right_type) {
+                ArrowDataType::UInt64
+            } else {
+                ArrowDataType::Float64
+            }
```

Comment:
Looks like this block can be extracted to a function. It's duplicated in `scalar_arith_op` too.